flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/deduplicate/DeduplicateFunctionHelper.java

```diff
@@ -147,6 +149,226 @@ static void processFirstRow(
 		out.collect(currentRow);
 	}
 
+	/**
+	 * Processes element to deduplicate on keys with row time semantic, sends current element if it is last row,
+	 * retracts previous element if needed.
+	 *
+	 * @param state                state of function
+	 * @param currentRow           latest row received by deduplicate function
+	 * @param out                  underlying collector
+	 * @param rowtimeIndex         index of row time field
+	 * @param generateUpdateBefore flag to generate UPDATE_BEFORE message or not
+	 * @param generateInsert       flag to gennerate INSERT message or not
+	 */
+	static void processLastRowOnRowtime(
+			ValueState<RowData> state,
+			RowData currentRow,
+			Collector<RowData> out,
+			int rowtimeIndex,
+			boolean generateUpdateBefore,
+			boolean generateInsert) throws Exception {
+
+		checkInsertOnly(currentRow);
+		RowData prevRow = state.value();
+		if (!isLastRow(prevRow, currentRow, rowtimeIndex)) {
+			return;
+		}
+		state.update(currentRow);
+
+		// store all needed data to state
+		collectRetractResult(
+				generateUpdateBefore,
+				generateInsert,
+				prevRow,
+				currentRow,
+				out,
+				null
+		);
+	}
+
+	/**
+	 * Processes element to deduplicate on keys with row time semantic, sends current element if it is last row,
+	 * retracts previous element if needed.
+	 *
+	 * @param state                state of function
+	 * @param bufferedRows         latest rows received by deduplicate function
+	 * @param serializer           serializer to serialize the data
+	 * @param out                  underlying collector
+	 * @param rowtimeIndex         index of row time field
+	 * @param generateUpdateBefore flag to generate UPDATE_BEFORE message or not
+	 * @param generateInsert       flag to gennerate INSERT message or not
+	 */
+	static void processMiniBatchLastRowOnRowtime(
+			ValueState<RowData> state,
+			List<RowData> bufferedRows,
+			TypeSerializer<RowData> serializer,
+			Collector<RowData> out,
+			int rowtimeIndex,
+			boolean generateUpdateBefore,
+			boolean generateInsert) throws Exception {
+
+		if (bufferedRows == null) {
+			return;
+		}
+
+		RowData preRow = state.value();
+		for (RowData currentRow : bufferedRows) {
+			checkInsertOnly(currentRow);
+			if (!isLastRow(preRow, currentRow, rowtimeIndex)) {
```

Comment:
The code is almost same with `processMiniBatchFirstRowOnRowtime`, we can abstract the common code out here.